[Add] Variations of local and shared algorithms

ckyiricky opened the pull request:
[Add] Add two variations of local-slice-algorithm
    1. Handle corner cases with a shared slice.
    2. Use weights to distribute 'float' endpoints
[Update] Score weights change
    Change weights for different metrics
    Add mean deviation and max deviation both into consideration

Commit:
[Update] Score weights change
Change weights for different metrics
Add mean deviation and max deviation both into consideration

Commit:
[Add/Update] Add/Updated three local variations
[Add]
Add one local variation that assigns the extra endpoints available at
the end to a global sliceGroup which has evenly distributed routing
weights to all zones
[Update]
Update local-weighted algorithms with clearer code and comments
Update local-shared with simpler implementation
    Treat urgent zones as a whole
    Use priority queue to assign endpoints to zones need endpoints
[Add/Update]
    Update input
    Update alg outputs and add new output for new algs

robscott commented on modeling/algorithm/algorithms.go:
```diff
@@ -27,6 +27,12 @@ func NewAlgorithm(name string) RoutingAlgorithm {
+	case "LocalWeighted":
+		klog.Info("LocalWeightedSliceAlgorithm created")
+		return LocalWeightedSliceAlgorithm{}
+	case "LocalOpt", "LocalAlgorithmOpt", "LocalSliceAlgorithmOpt":
```
Maybe `LocalShared` instead to match the type.

ckyiricky replied:
Fixed, sorry for the silly mismatch here.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+// LocalSharedSliceAlgorithm is one variation of LocalSliceAlgorithm which
+// 'borrows' and 'rents' endpoints from other zones to make the local
+// EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution). This variation deals with failed corner cases while having a
```
How does it deal with failed corner cases?

ckyiricky replied:
Added more comments about it

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy, zones with 0.xx expected endpoints while have no endpoints in
+// zone will be treated with a higher priority.
```
Maybe something like this?
```suggestion
// zone' policy, zones with no endpoints allocated will be treated with a higher priority.
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+	}
+	// if number of total endpoints < number of zones, use original algorithm
+	// instead. This algorithm itself can handle some of these special corner
+	// cases but performs poorly in most of the time, so using the original
```
```suggestion
	// cases but performs poorly at larger scale, so using the original
```

ckyiricky replied:
change to more specific cases

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+	endpointsNeeded := endpointsList{}
+	// endpointsNeeded stores high priority zones with number of endpoints
+	// needed
+	endpointsNeededUrgent := endpointsList{}
```
```suggestion
	// endpointsNeededUrgent stores high priority zones with number of endpoints
	// needed
	endpointsNeededUrgent := endpointsList{}
```

ckyiricky replied:
done

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+	heap.Init(&urgentPool)
+	heap.Init(&receiverPool)
+
+	succ, err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &urgentPool, &receiverPool)
```
Since we know that we don't have to deal with scenarios where the number of endpoints is less than the number of zones, maybe we can simplify this a bit. This is likely an oversimplification, but it seems like we want to treat the case of zones with no local endpoints as a special one. In that case, we would simply have a single slice group that contained `max(sum(expectedEndpoints), 1)` endpoints and was distributed to all zones that didn't have local endpoints. The rest would then be a relatively simple assignment problem. Admittedly I'm likely missing some of the edge cases here.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,254 @@
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints. It serves
+// high-priority zones first.
+func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, endpointsNeededUrgent *endpointsList, sliceGroups map[string]types.EndpointSliceGroup, urgentPool *ZonePriorityQueue, receiverPool *ZonePriorityQueue) (bool, error) {
```
I think we may be able to simplify this to avoid the need for urgent special cases.

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,202 @@
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means need more endpoints from other
+		// sliceGroups, a positive value means need give out endpoints to other
+		// sliceGroups
```
```suggestion
		// deviation: a negative value means this zone needs more endpoints from other
		// zones, a positive value means this zone needs to give out endpoints to other
		// zones
```

ckyiricky replied:
Done

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,202 @@
+		// push 0.xx deviation into corresponding lists
+		floatDeviation := deviation - float64(intDeviation)
+		if floatDeviation > 0 {
+			// Intentionally set deviation to default 0, this means the zone
+			// itself needs float part from this endpoint otherwise means the
+			// endpoints will be shared by other zones only
```
This comment is a bit confusing to me

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,202 @@
+			// endpoints will be shared by other zones only
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, weight: floatDeviation})
+		} else if floatDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -floatDeviation})
```
Why is deviation 1 here? Why is weight -floatDeviation?

ckyiricky replied:
Done

Commit:
[Add] Add shared-global-algorithm variation
Add variation that excludes contributor reaching global slices

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,202 @@
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
+
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// No more 'int' endpoints available but are still 'int' endpoints in
+		// need, we push the needed entpoints into weighted list and deal them
+		// as float endpoints
```
Not quite sure I'm understanding this correctly, but maybe something like:
```
There are no more full endpoints available, but this zone still needs endpoints. Push the needed endpoints into a weighted list and deal with them as partial endpoints.
```

ckyiricky replied:
done

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,202 @@
+				continue
+			}
+		}
+		// if needed.deviation != 0 means more 'int' needed than available, push
+		// to weighted list and deal as float
```
Isn't this similar to the comment on 103?

ckyiricky replied:
yes, but this is an edge case that the '103' can't deal with. Simply its one deviation has remaining number needed but no more available endpoints to serve it, we have to push that deviation into float list and treat as float endpoints.

robscott reviewed (commented):
Thanks for the work on this! Didn't quite make it through everything for this first round, but had some higher level questions around structure.

ckyiricky commented:
[Add/Update] Add/Updated three local variations
    [Add]
    Add one local variation that assigns the extra endpoints available at
    the end to a global sliceGroup which has evenly distributed routing
    weights to all zones
    [Update]
    Update local-weighted algorithms with clearer code and comments
    Update local-shared with simpler implementation
        Treat urgent zones as a whole
        Use priority queue to assign endpoints to zones need endpoints
    [Add/Update]
        Update input
        Update alg outputs and add new output for new algs

ckyiricky commented:
Add a shared global variation which excludes contributors from the global slices

robscott commented on example/topology-sim-input.csv:
```diff
@@ -5,6 +5,7 @@ name,zone1,zone2,zone3
 balanced minimal endpoints,30 1,30 1,30 1
 unbalanced minimal endpoints,30 2,30 1,30 0
 only 1 endpoint,30 1,30 0,30 0
+only 2 endpoint,30 2,30 0,30 0
```
I think it would be good to also add one or more of the edge cases we'd discussed earlier. Can't quite remember exactly what they were, but I know there was one something like this:

```
a b c
2 3, 0 1, 0 1 
```

ckyiricky replied:
Created an issue for the corner cases update

robscott commented on modeling/algorithm/algorithms.go:
```diff
@@ -23,10 +23,22 @@ func NewAlgorithm(name string) RoutingAlgorithm {
-		return SharedGlobalAlgorithm{globalWeight: 0.4, globalThreshold: 100}
+		return SharedGlobalAlgorithm{sharedCoreAlgorithm: SharedGlobalAlgorithmCore{globalWeight: 0.4, globalThreshold: 100}}
+	case "SharedGlobalExclude", "SharedGlobalAlgorithmExclude":
+		klog.Info("SharedGlobalAlgorithmExclude created")
```
I think this should by `SharedMultiZone`

ckyiricky replied:
changed

robscott commented on modeling/algorithm/algorithms.go:
```diff
@@ -23,10 +23,22 @@ func NewAlgorithm(name string) RoutingAlgorithm {
+	case "LocalOpt", "LocalAlgorithmOpt", "LocalSliceAlgorithmOpt":
+		klog.Info("LocalSliceAlgorithmOpt created")
+		return LocalSliceAlgorithmOpt{}
+	case "LocalShared":
```
Nit: For some algorithms you have multiple acceptable name and for others you don't. I think for consistency it would be best to support a single name per algorithm.

ckyiricky replied:
That makes sense, I changed to two names for every algorithms : xxx and xxxAlgorithm

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+
+// LocalSharedSliceAlgorithm is one variation of LocalSliceAlgorithm which
+// 'borrows' and 'rents' endpoints from other zones to make the local
+// EndpointSlice balanced with the incoming traffic (number of nodes
```
Nit: EndpointSliceGroup, EndpointSlices or similar to indicate it isn't just a single EndpointSlice.

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+// 'borrows' and 'rents' endpoints from other zones to make the local
+// EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution). This variation deals with failed corner cases by sharing
+// endpoints to zones has no endpoints.
```
Nit: that have no endpoints

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+	// original algorithm seems a better solution in terms of performance and
+	// simplicity.
+	if region.TotalEndpoints < len(region.ZoneDetails) {
+		return OriginalAlgorithm{}.CreateSliceGroups(region)
```
I forget, is the OrignalAlgorithm just the global/shared one? If so, maybe we should change the name here to make it more clear what it does.

ckyiricky replied:
original algorithms if the  'evenly distributed' policy.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+	sliceGroups := map[string]types.EndpointSliceGroup{}
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	// endpointsNeededUrgent stores high priority zones with number of endpoints
```
Maybe explain what high priority zones are here

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+			availablePool.ZoneNames = append(availablePool.ZoneNames, zoneName)
+		}
+		receiverPool.ZoneNames = append(receiverPool.ZoneNames, zoneName)
+		// merge all the zones have no endpoints into a whole
```
Maybe something like this?
```suggestion
		// merge all the zones with no endpoints into a shared slice group
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+		receiverPool.ZoneNames = append(receiverPool.ZoneNames, zoneName)
+		// merge all the zones have no endpoints into a whole
+		if zone.Endpoints == 0 {
+			// this is a form used to accuratly represent the deviation in
```
```suggestion
			// this is a form used to accurately represent the deviation in
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+		// merge all the zones have no endpoints into a whole
+		if zone.Endpoints == 0 {
+			// this is a form used to accuratly represent the deviation in
+			// float, floatDeviation = 1 * -deviation
```
This comment is confusing, did you mean to remove it?

ckyiricky replied:
add more comments to explain the move here

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+	heap.Init(&availablePool)
+	heap.Init(&receiverPool)
+
+	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &availablePool, &receiverPool)
```
Not quite sure I understand this, why do availablePool and receiverPool get equivalent slice groups assigned to them? Can any of this logic be moved into balanceSliceGroups?

ckyiricky replied:
move into balanceSliceGroups

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+
+	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &availablePool, &receiverPool)
+	if !succ {
+		klog.Infoln("fail to do local algorithm, switch to shared global algorithm")
```
```suggestion
		klog.Infoln("failed to use local algorithm, switching to shared global algorithm")
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+	// merge one sharedSG that zones in the urgent list will consume
+	mergedSG := types.EndpointSliceGroup{Composition: map[string]types.WeightedEndpoints{}, ZoneTrafficWeights: map[string]float64{}}
+	mergedED := endpointDeviation{name: "merged"}
+	mergedDeviation := 0.0
```
I know it's hard to name similar things here, but maybe a comment or something would help explain what the difference between these 2 deviations is.

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+		mergedSG.ZoneTrafficWeights[urgentZone.name] = 1
+		endpointsNeededUrgent.pop()
+	}
+	// take the ceil, in case the deviation = 0.xx, we have to make sure there
```
Would this also work?
```suggestion
	// take the ceil, if the deviation > 0, we have to make sure there
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+		}
+		candidate := heap.Pop(availablePool).(string)
+		expectedEndpoints := float64(availablePool.Region.TotalEndpoints) * availablePool.Region.ZoneDetails[candidate].NodesRatio
+		if float64(sliceGroups[candidate].Composition[candidate].Number)-expectedEndpoints >= 1 {
```
There's a lot going on this if statement - could use some nil checks. Maybe this could be a helper func.

ckyiricky replied:
added a helper function

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+			break
+		}
+		candidate := heap.Pop(availablePool).(string)
+		expectedEndpoints := float64(availablePool.Region.TotalEndpoints) * availablePool.Region.ZoneDetails[candidate].NodesRatio
```
This could use a nil check, maybe a helper func would be good here

ckyiricky replied:
added a nil detection inside the helper function. In this case, it should be guaranteed to be non-nil.

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+			heap.Push(receiverPool, receiveZone)
+
+			updateSGComposition(sliceGroups[candidate], candidate, -1, 1)
+			heap.Push(availablePool, candidate)
```
There's a lot of key logic here, some comments to explain what is happening in each step would be helpful

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-shared-algorithm.go:
```diff
@@ -0,0 +1,205 @@
+			heap.Push(availablePool, candidate)
+			continue
+		}
+		break
```
This continue and break combination along with an otherwise infinite loop feels a bit strange to me, is there another way we could structure this loop?

ckyiricky replied:
restructured

robscott commented on modeling/algorithm/local-slice-algorithm-opt.go:
```diff
@@ -0,0 +1,136 @@
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: int(-deviation)})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(zone.Endpoints), Weight: 1}
+		} else {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(zone.Endpoints), Weight: 1}
```
This seems to be almost completely repeated in each of the three clauses above, maybe initialize the weightedEndpoints above and only change the number in the one case it's necessary.

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-slice-algorithm-opt.go:
```diff
@@ -0,0 +1,136 @@
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: sendZone.deviation, Weight: 1}
+				receiveZone.deviation -= sendZone.deviation
+				endpointsAvailable.pop()
+				continue
```
Do we need continue here?

ckyiricky replied:
deleted

robscott commented on modeling/algorithm/local-slice-algorithm-opt.go:
```diff
@@ -0,0 +1,136 @@
+	// endpoints in need
+	if len(endpointsAvailable.byZone) != 0 {
+		// in this case, we assign those extra endpoints to a global
+		// endpointSliceGroups
```
```suggestion
		// endpointSliceGroup
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-slice-algorithm-opt.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	// all endpoints should have been distributed properly. This happens when
+	// the sum of approximated available endpoints > sum of approximated
+	// endpoints in need
```
These two sentences seem to contradict each other. If all endpoints had been distributed at this point it seems like we wouldn't need this section.

ckyiricky replied:
Deleted the first sentence

robscott commented on modeling/algorithm/local-slice-types.go:
```diff
@@ -0,0 +1,118 @@
+/*
```
This seems to include more than just local slice types, maybe a more generic utils.go name would be a better fit here.

ckyiricky replied:
renamed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+		// other zones, a positive value means this zone needs to give out
+		// endpoints to other zones
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// intDeviation is dealt same way as original local-slice algorithm that
```
```suggestion
		// intDeviation is dealt with the same way as original local-slice algorithm that
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
```
I had a similar comment above that these WeightedEndpoints could likely be declared above the if statements to reduce the duplication.

ckyiricky replied:
changed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+		// endpoints and 2 endpoints needed
+		decimalDeviation := deviation - float64(intDeviation)
+		if decimalDeviation > 0 {
+			// One endpoint from this zone will contribute 1-decimalDeviation to
```
What's the logic behind this equation, maybe add a quick explanation of why this equation is used here as a comment.

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+			// will be implemented through routing weights of EndpointSliceGroup
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, deviation: 1, weight: 1 - decimalDeviation, consumeByLocal: true})
+		} else if decimalDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -decimalDeviation})
```
Maybe add a comment here as well to explain why weight is being set to -decimalDeviation.

ckyiricky replied:
added

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
```
How similar is this to the other balanceSliceGroups functions? Is there a way we could share some code here between algorithms?

ckyiricky replied:
Created a helper function to share some common logic

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+			endpointsNeeded.pop()
+			continue
+		}
+		// same logic as original local-slice algorithm
```
If this is shared logic maybe we can extract it to a helper func?

ckyiricky replied:
wrapped the logic to a helper function, same as the previous comment

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+			}
+		}
+		// if needed.deviation != 0 means more full endpoints needed than
+		// available, push to weighted list and deal them as partial endpoints
```
```suggestion
		// available, push to weighted list and deal with them as partial endpoints
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+				continue
+			}
+		}
+		// if needed.deviation != 0 means more full endpoints needed than
```
When would this happen? It seems strange to equate != 0 meaning more needed than available, wouldn't a negative value have a different meaning than a positive value here?

ckyiricky replied:
changed to > 0 and added more comments to explain

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+	// weight to zoneA
+	for _, extraEndpoints := range weightedEndpointsAvailable.byZone {
+		sharedSlice := types.EndpointSliceGroup{Label: "shared", Composition: map[string]types.WeightedEndpoints{}, ZoneTrafficWeights: map[string]float64{}}
+		// If this endpoint will be consumbed by its local zone, contribue to
```
```suggestion
		// If this endpoint will be consumed by its local zone, contribute to
```

ckyiricky replied:
fixed

robscott commented on modeling/algorithm/local-weighted-slice-algorithm.go:
```diff
@@ -0,0 +1,211 @@
+				sharedSlice.Label += "-" + receiveZone.name
+				extraEndpoints.weight -= sharedSlice.ZoneTrafficWeights[receiveZone.name]
+				weightedEndpointsNeeded.pop()
+				continue
```
Wouldn't this happen automatically?

ckyiricky replied:
deleted

robscott commented on modeling/algorithm/shared-global-algorithm-core.go:
```diff
@@ -0,0 +1,101 @@
+// EndpointSliceGroups
+func (alg SharedGlobalAlgorithmCore) CreateSliceGroups(region types.RegionInfo, excludeContributor bool) (map[string]types.EndpointSliceGroup, error) {
+	if region.ZoneDetails == nil {
+		return nil, errors.New("can't create EndpointSlices with 0 number of zone")
```
```suggestion
		return nil, errors.New("can't create EndpointSlices without zones specified")
```

ckyiricky replied:
changed

robscott commented on modeling/algorithm/shared-global-algorithm-core.go:
```diff
@@ -0,0 +1,101 @@
+
+	// Output EndpointSlices
+	sliceGroups := make(map[string]types.EndpointSliceGroup)
+	// The global sliceGroup -- might be split into many small global slices
```
Had a similar question above, what does `--` mean here?

ckyiricky replied:
sorry for the legacy comment here, updated to a more reasonable one

robscott commented on modeling/algorithm/shared-global-algorithm-core.go:
```diff
@@ -0,0 +1,101 @@
+	globalSliceGroup.ZoneTrafficWeights = make(map[string]float64)
+	for name, zone := range region.ZoneDetails {
+		var globalEndpoints types.WeightedEndpoints
+		// If total pods <= threshold, all pods go to global slice
+		if region.TotalEndpoints <= alg.globalThreshold {
+			globalEndpoints.Number = zone.Endpoints
+			globalEndpoints.Weight = 1
```
Elsewhere you're just using the original algorithm in similar cases, would that make sense here as well?

ckyiricky replied:
sorry about the legacy issue here, updated to using original algorithm

robscott reviewed (commented):
Thanks for all your work on this, lots of good stuff in here! Really appreciate all the comments throughout to help explain the logic. One overarching comment is that it seems like there's a lot of duplicate code between these algorithms. I know relying on shared functions could overcomplicate changes to a specific algorithm in the future, but maybe there's some room to share some code that's less likely to change.

Commit:
[Update] variation update
Update local* algorithms with fewer duplicated codes
Rename shared-global-algorithm-exclude to shared-multizone-algorithm
Fix typos

ckyiricky commented:
[Update] variation update
    Update local* algorithms with fewer duplicated codes
    Rename shared-global-algorithm-exclude to shared-multizone-algorithm
    Fix typos

robscott reviewed (approved):
Thanks for all your work on this!
